// plugin-typescript/tmp/es6/transpiler.js
import ts from 'typescript';
import { isJavaScript, isJSX, isSourceMap, hasError } from './utils';
import Logger from './logger';
var logger = new Logger({ debug: false });
var Transpiler = (function () {
    function Transpiler(_host) {
        this._host = _host;
    }
    Transpiler.prototype.getTranspileOptions = function (options) {
        var result = ts.clone(options);
        result.isolatedModules = true;
        if (result.sourceMap === undefined)
            result.sourceMap = result.inlineSourceMap;
        if (result.sourceMap === undefined)
            result.sourceMap = true;
        result.inlineSourceMap = false;
        result.declaration = false;
        result.noEmitOnError = false;
        result.out = undefined;
        result.outFile = undefined;
        result.noLib = true;
        result.lib = undefined;
        result.types = [];
        result.suppressOutputPathCheck = true;
        result.noEmit = false;
        return result;
    };
    Transpiler.prototype.transpile = function (sourceName, options) {
        logger.debug("transpiling " + sourceName);
        var file = this._host.getSourceFile(sourceName);
        if (!file)
            throw new Error("file [" + sourceName + "] has not been added");
        if (!file.output) {
            var transpileOptions = this.getTranspileOptions(options);
            var program = ts.createProgram([sourceName], transpileOptions, this._host);
            var jstext_1 = undefined;
            var maptext_1 = undefined;
            var emitResult = program.emit(undefined, function (outputName, output) {
                if (isJavaScript(outputName) || isJSX(outputName))
                    jstext_1 = output.slice(0, output.lastIndexOf("//#"));
                else if (isSourceMap(outputName))
                    maptext_1 = output;
                else
                    throw new Error("unexpected ouput file " + outputName);
            });
            var diagnostics = emitResult.diagnostics
                .concat(program.getOptionsDiagnostics())
                .concat(program.getSyntacticDiagnostics());
            file.output = {
                failure: hasError(diagnostics),
                diags: diagnostics,
                js: jstext_1,
                sourceMap: maptext_1
            };
        }
        return file.output;
    };
    return Transpiler;
}());
export { Transpiler };
